src/utils/infra_link_selector.cpp

```diff
@@ -0,0 +1,264 @@
+ */
+
+#if __linux__
+
+#define OTBR_LOG_TAG "ILS"
+
+#include "utils/infra_link_selector.hpp"
+
+#include <linux/rtnetlink.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "utils/socket_utils.hpp"
+
+namespace otbr {
+namespace Utils {
+
+constexpr std::chrono::milliseconds InfraLinkSelector::kInfraLinkSelectionDelay;
+
+InfraLinkSelector::InfraLinkSelector(std::vector<const char *> aInfraLinkNames)
+    : mInfraLinkNames(std::move(aInfraLinkNames))
+{
+    if (mInfraLinkNames.size() >= 2)
+    {
+        mNetlinkSocket = CreateNetLinkRouteSocket(RTMGRP_LINK);
+        VerifyOrDie(mNetlinkSocket != -1, "Failed to create netlink socket");
+    }
+}
+
+InfraLinkSelector::~InfraLinkSelector(void)
+{
+    if (mNetlinkSocket != -1)
+    {
+        close(mNetlinkSocket);
+        mNetlinkSocket = -1;
+    }
+}
+
+const char *InfraLinkSelector::Select(void)
+{
+    const char *   oldInfraLink = mCurrentInfraLink;
+    auto           now          = Clock::now();
+    constexpr auto kMinTime     = Clock::time_point::min();
+
+    VerifyOrExit(!mInfraLinkNames.empty(), mCurrentInfraLink = kDefaultInfraLinkName);
+    VerifyOrExit(mInfraLinkNames.size() > 1, mCurrentInfraLink = mInfraLinkNames.front());
+    VerifyOrExit(mRequireReselect, assert(mCurrentInfraLink != nullptr));
+
+    if (mCurrentInfraLink != nullptr)
+    {
+        // Prefer `mCurrentInfraLink` if it's up and running.
+        if (IsInfraLinkRunning(mCurrentInfraLink) == kUpAndRunning)
+        {
+            mCurrentInfraLinkDownTime = kMinTime;
+            ExitNow();
+        }
+
+        if (mCurrentInfraLinkDownTime == kMinTime)
+        {
+            mCurrentInfraLinkDownTime = now;
+        }
+
+        // Prefer `mCurrentInfraLink` if it's down for less than `kInfraLinkSelectionDelay`
+        if (now - mCurrentInfraLinkDownTime < kInfraLinkSelectionDelay)
+        {
+            std::chrono::milliseconds delay =
+                kInfraLinkSelectionDelay -
+                std::chrono::duration_cast<std::chrono::milliseconds>(now - mCurrentInfraLinkDownTime);
+
+            mTaskRunner.Post(delay, [this]() { mRequireReselect = true; });
+            ExitNow();
+        }
+    }
+
+    {
+        const char *bestInfraLink;
+        LinkState   bestState;
+
+        EvaluateInfraLinks(bestInfraLink, bestState);
+
+        // Prefer `mCurrentInfraLink` if no other infra link is up and running
+        if (mCurrentInfraLink != nullptr && bestState != kUpAndRunning)
+        {
+            ExitNow();
+        }
+
+        // Prefer the infra link in the best state
+        assert(mCurrentInfraLink != bestInfraLink);
+
+        mCurrentInfraLink         = bestInfraLink;
+        mCurrentInfraLinkDownTime = bestState == kUpAndRunning ? kMinTime : now;
+    }
+
+exit:
+    mRequireReselect = false;
+
+    if (oldInfraLink != mCurrentInfraLink)
```

Comment:
Note that you're comparing two pointers.

Reply:
It needs to be changed if we are refactoring the PR to use std::string.